common/network-common/src/main/java/org/apache/spark/network/sasl/aes/AesCipher.java

```diff
@@ -0,0 +1,298 @@
+  private final IvParameterSpec outIvSpec;
+  private Properties properties;
+
+  public AesCipher(
+      Properties properties,
+      byte[] inKey,
+      byte[] outKey,
+      byte[] inIv,
+      byte[] outIv) throws IOException {
+    this.properties = properties;
+    inKeySpec = new SecretKeySpec(inKey, "AES");
+    inIvSpec = new IvParameterSpec(inIv);
+    outKeySpec = new SecretKeySpec(outKey, "AES");
+    outIvSpec = new IvParameterSpec(outIv);
+  }
+
+  public AesCipher(AesConfigMessage configMessage) throws IOException  {
+    this(new Properties(), configMessage.inKey, configMessage.outKey,
+      configMessage.inIv, configMessage.outIv);
+  }
+
+  /**
+   * Create AES crypto output stream
+   * @param ch The underlying channel to write out.
+   * @return Return output crypto stream for encryption.
+   * @throws IOException
+   */
+  private CryptoOutputStream createOutputStream(WritableByteChannel ch) throws IOException {
+    return new CryptoOutputStream(TRANSFORM, properties, ch, outKeySpec, outIvSpec);
+  }
+
+  /**
+   * Create AES crypto input stream
+   * @param ch The underlying channel used to read data.
+   * @return Return input crypto stream for decryption.
+   * @throws IOException
+   */
+  private CryptoInputStream createInputStream(ReadableByteChannel ch) throws IOException {
+    return new CryptoInputStream(TRANSFORM, properties, ch, inKeySpec, inIvSpec);
+  }
+
+  /**
+   * Add handlers to channel
+   * @param ch the channel for adding handlers
+   * @throws IOException
+   */
+  public void addToChannel(Channel ch) throws IOException {
+    ch.pipeline()
+      .addFirst(ENCRYPTION_HANDLER_NAME, new AesEncryptHandler(this))
+      .addFirst(DECRYPTION_HANDLER_NAME, new AesDecryptHandler(this));
+  }
+
+  /**
+   * Create the configuration message
+   * @param conf is the local transport configuration.
+   * @return Config message for sending.
+   */
+  public static AesConfigMessage createConfigMessage(TransportConf conf) {
+    int keySize = conf.aesCipherKeySize();
+    Properties properties = getProperties(conf);
+
+    try {
+      int paramLen = CryptoCipherFactory.getCryptoCipher(AesCipher.TRANSFORM, properties)
+        .getBlockSize();
+      byte[] inKey = new byte[keySize];
+      byte[] outKey = new byte[keySize];
+      byte[] inIv = new byte[paramLen];
+      byte[] outIv = new byte[paramLen];
+
+      CryptoRandom random = CryptoRandomFactory.getCryptoRandom(properties);
+      random.nextBytes(inKey);
+      random.nextBytes(outKey);
+      random.nextBytes(inIv);
+      random.nextBytes(outIv);
+
+      return new AesConfigMessage(keySize, inKey, inIv, outKey, outIv);
+    } catch (Exception e) {
+      logger.error("AES config error", e);
+      throw Throwables.propagate(e);
+    }
+  }
+
+  private static Properties getProperties(TransportConf conf) {
+    Properties props = new Properties();
+    props.setProperty(CryptoCipherFactory.CLASSES_KEY, conf.aesCipherClass());
```

Comment:
You're not adding configs for `CryptoRandomFactory` here. Really what you want here is `CryptoStreamUtils.toCryptoConf`, but `TransportConf` doesn't really have the interface to support that. So at least add a TODO so that we know that this needs to be done.